### PR TITLE
[FLINK-10884] [yarn/mesos]  adjust  container memory param  to set a safe margin from offheap memory

### PR DESCRIPTION
**Pull request by wg1026688210**
What is the purpose of the change
This pull request adjust yarn container memory parameters/

Brief change log
1.set a safe margin  from offHeapMemory .
2.add a container memory config  "containerized.offheap-cutoff-ratio" default 0.25f   at org.apache.flink.configuration.TaskManagerOptions  
3. change the unit test 
Verifying this change
This change is a trivial rework / code cleanup without any test coverage.

Does this pull request potentially affect one of the following parts:
Dependencies (does it add or upgrade a dependency):  no
The public API, i.e., is any changed class annotated with @Public(Evolving): no
The serializers: no
The runtime per-record code paths (performance sensitive): no
Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: yes
The S3 file system connector: i don't know
Documentation
Does this pull request introduce a new feature? add a config
If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented) not documented

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/ContaineredTaskManagerParameters.java [hunk 1]
```diff
@@ -158,8 +158,10 @@ public static ContaineredTaskManagerParameters create(
 
 		// (2) split the remaining Java memory between heap and off-heap
 		final long heapSizeMB = TaskManagerServices.calculateHeapSizeMB(containerMemoryMB - cutoffMB, config);
-		// use the cut-off memory for off-heap (that was its intention)
-		final long offHeapSizeMB = containerMemoryMB - heapSizeMB;
+		// (3) try to compute the offHeapMemory from a safe margin
+		final long restMemoryMB = containerMemoryMB - heapSizeMB;
+		final long offHeapCutoffMemory = calculateOffHeapCutoffMB(config, restMemoryMB);
```
[the remark this answers is not in the document]
Comment (wg1026688210) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/ContaineredTaskManagerParameters.java]: Thank you for your attention. I do add a containerized.offheap-cutoff-ratio for two reason
 1.  it will be unfriendly for  a small memory container such as 1gb memory container  if we only use containerized.heap-cutoff-ratio and containerized.heap-cutoff-min to cutoff the memory 0f container physical memory .  
 2. native memory  not only  contains memory which used by rockdb state backend but also   memory pools are called arenas which will cost enormous memory in multi-core cpu machine
as time goes on . You can use 'pmap -x {TM container pid}' command to validate it 
 So I want to  add a  option to  adjust memory  flexiblely  for someone who wana adjust both heap meomory and offheap memory to adapt the three kinds of memory.

[the remark this answers is not in the document]
Comment (wg1026688210) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/ContaineredTaskManagerParameters.java]: Thanks for your suggestion.  It maybe  bring  more refactor  to the project  if  we name the current "containerized.heap-cutoff-ratio" to "containerized.memory-cutoff-ratio".Because we need to adapt  a small memory container due to 600 mega byte of the current "containerized.heap-cutoff-min". How about we set a  common default value of "containerized.offheap-cutoff-ratio" and "containerized.offheap-cutoff-min" so that user do not need to care the two parameters at most condition.

[the remark this answers is not in the document]
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/ContaineredTaskManagerParameters.java]: Sorry I have not got the point of more refactor you mentioned. I think you just need to adjust both parameters of `containerized.memory-cutoff-ratio` and `containerized.memory-cutoff-min` for your project. 

Unless in your case, modify the existing parameters are not very convenient, and you want to add extra two parameters  to avoid affecting existing ones.

I think it is better to let another related committer check this option.  @GJL what do you think?

[the remark this answers is not in the document]
Comment (wg1026688210) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/ContaineredTaskManagerParameters.java]: maybe we need some suggestions from @GJL

**Comment by NicoK** [comment]
Let me drop some general notes: the `cutoff` was meant to include anything but the heap and thus cover for classes (metaspace), off-heap, etc. and when I touched this code, I kept that.
If you look into https://github.com/apache/flink/blob/master/flink-dist/src/main/flink-bin/bin/taskmanager.sh#L65, you'll also see that we practically set it to infinity there. Having an actual limit (as is used through `ContaineredTaskManagerParameters`) will only result in a different error in case you are hitting it: `java.lang.OutOfMemoryError` instead of being killed by YARN/Mesos/etc. That may be better for diagnosing it but doesn't solve the root cause of something using up too much memory.

We also need to touch all memory calculations for https://issues.apache.org/jira/browse/FLINK-10317 again - maybe splitting this off is already enough for you? So having `totalContainerMem = heap + offHeap + Metaspace`.

**Comment by GJL** [comment]
Sorry, I am very late to the game. 

For YARN TaskManager containers, _-XX:MaxDirectMemorySize_ is set to `offHeapSizeMB`: https://github.com/apache/flink/blob/3fba21a0ce5534b8c95a2bb9e8547b0061e5bb8f/flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/BootstrapTools.java#L430-L433

We could introduce an offheap cutoff. However, I think @NicoK's comment addresses a valid point here:

> Having an actual limit (as is used through ContaineredTaskManagerParameters) will only result in a different error in case you are hitting it: java.lang.OutOfMemoryError instead of being killed by YARN/Mesos/etc. That may be better for diagnosing it but doesn't solve the root cause of something using up too much memory.

The bottom line is that unless the TM's offheap memory allocations respect _-XX: MaxDirectMemorySize_, you will probably run into another error. Please correct me if I am wrong. 

@wg1026688210:
Did you try out your changes in production, and did it solve your problem?
Were you able to work around your problem otherwise, e.g., by increasing the total container memory?

**Comment by GJL** [comment]
I am closing this PR for now. Feel free to reopen if needed.

**Comment by wg1026688210** [comment]
thanks for reply  I learn many thing from this discuss
